File: 1-TwoSum/main.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i, v := range nums {
		c := target - v 
		if j, ok := m[c]; ok {
			return []int{j, i}
		}
		
		m[v] = i
	}

	return nil 
}

func case1() ([]int, int) {
	nums := []int{2, 7, 11, 12}
	target := 9
	return nums, target
}

func case2() ([]int, int) {
	nums := []int{3, 2, 4}
	target := 6
	return nums, target
}

func case3() ([]int, int) {
	nums := []int{3, 3}
	target := 6
	return nums,target
}

func main() {
	fmt.Println(twoSum(case1()))
	fmt.Println(twoSum(case2()))
	fmt.Println(twoSum(case3()))
}

File: 1-TwoSum/Main.java
import java.util.HashMap;
import java.util.Map;

class TwoSum {
  public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> m = new HashMap<>();

    for (int i = 0; i < nums.length; i++) {
      int complement = target - nums[i];
      if (m.containsKey(complement)) {
        return new int[] { m.get(complement), i };
      }
      m.put(nums[i], i);
    }

    return new int[0];
  }

  public static void main(String[] args) {
    TwoSum sol = new TwoSum();
    int[] result = sol.twoSum(new int[] { 2, 7, 11, 12 }, 9);
    System.out.println("[" + result[0] + ", " + result[1] + "]");
  }
}
